File: frontend/machsense/lib/api/telemetry.ts
import { api } from "./client";

export interface TelemetryOut {
  id: number;
  machine_id: number;
  timestamp?: string;
  rpm?: number;
  temperature?: number;
  vibration_rms?: number;
  kurtosis?: number;
  crest_factor?: number;
  dominant_frequency?: number;
  health_score?: number;
}

export interface TelemetryCreate {
  machine_id: number;
  rpm?: number;
  temperature?: number;
  accel_x?: number;
  accel_y?: number;
  accel_z?: number;
  vibration_rms?: number;
  kurtosis?: number;
  crest_factor?: number;
  dominant_frequency?: number;
  health_score?: number;
}

export const telemetryApi = {
  getRecent: (machine_id: number, limit = 60) =>
    api.get<TelemetryOut[]>(`/telemetry/${machine_id}?limit=${limit}`),
  ingest: (data: TelemetryCreate) =>
    api.post<TelemetryOut>("/telemetry/ingest", data),
  injectMock: (machine_id: number, mode: "normal" | "warning" | "critical" = "normal") =>
    api.post<TelemetryOut>(`/telemetry/mock/${machine_id}?mode=${mode}`),
};

File: backend/app/db/models/telemetry.py
from sqlalchemy import Column, Integer, Float, DateTime, ForeignKey, func
from app.db.database import Base


class Telemetry(Base):
    __tablename__ = "telemetry"

    id = Column(Integer, primary_key=True, index=True)
    machine_id = Column(Integer, ForeignKey("machines.id"), nullable=False, index=True)
    timestamp = Column(DateTime(timezone=True), server_default=func.now(), index=True)

    rpm = Column(Float, nullable=True)
    temperature = Column(Float, nullable=True)
    accel_x = Column(Float, nullable=True)
    accel_y = Column(Float, nullable=True)
    accel_z = Column(Float, nullable=True)
    vibration_rms = Column(Float, nullable=True)
    kurtosis = Column(Float, nullable=True)
    crest_factor = Column(Float, nullable=True)
    dominant_frequency = Column(Float, nullable=True)
    health_score = Column(Float, nullable=True)
